### PR TITLE
Immediately invalidate on lost leadership

### PR DESCRIPTION
**Pull request by schlosna**
Comments on #6824 , expanding AwaitingLeadershipProxy tests for various scenarios

### CODE REVIEW

File: leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java
Comment (jeremyk-91) [inline leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java:167]: thanks for fixing this!

**Review by jeremyk-91 — COMMENTED** [review]
I believe this makes sense, though as I'll be out for a bit, handing over to @LucasIME or @mdaudali to verify things are OK before merging

File: leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java
Comment (mdaudali) [inline leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java:116]: I don't believe this is correct. Whilst the return type is the `NotCurrentLeaderException`, this method does not return the exception, but throws at evaluation. This means that instead of returning a future that fails, the handleInvocation method itself throws the exception, which is not what's intended here.

I wrote a mini test: 
```
    @Test
    public void areTheyTheSame() {
        Future<?> future = Futures.immediateFailedFuture(returnException());
        assertThat(future).isDone();
    }

    RuntimeException returnException() {
        return new RuntimeException("This is a test");
    }

    RuntimeException throwException() {
        throw new RuntimeException("This is a test");
    }
```

using returnException, the test passes. With throwException, the test fails as the method threw an exception, rather than the future failing

Comment (mdaudali) [reply leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java:116]: I'll admit I'm a little surprised / concerned that none of our tests caught this.

File: leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java
Comment (mdaudali) [inline leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java]: Thanks! Would you mind leaving a comment + adding a test that covers this code path when thrown a NotCurrentLeaderException? Just want to make sure we don't regress if we change the parent type of the exception - and I couldn't see one where we had a test covering that...

**Review by mdaudali — CHANGES_REQUESTED** [review]
Thanks for the change! I don't believe, in its current form, that it works as intended, but do correct me if I'm wrong

File: leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java
Comment (schlosna) [reply leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java:116]: good catch, updated and expanded test coverage

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosTimestampBoundStore.java
Comment (mdaudali) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosTimestampBoundStore.java]: Any reason you chose to change the return type but not change the semantics to return the exception instead?

File: leader-election-impl/src/test/java/com/palantir/leader/proxy/AwaitingLeadershipProxyTest.java
Comment (mdaudali) [inline leader-election-impl/src/test/java/com/palantir/leader/proxy/AwaitingLeadershipProxyTest.java:197]: Thanks for adding the tests! We're still not testing the case where we throw rather than returning a failed future (i.e., the mistake in the previous iteration), because the handleInvocation method unwraps the future anyway if the method under the proxy does not return a ListenableFuture. As a result, in the current test framework, the cases of throwing vs returning a failed future are not visibly different.

That said, from a review perspective, I'm happy with the change. If you want to go ahead with trying to get our tests to repro the above failure mode, please feel free and I can give it another review.

**Review by mdaudali — APPROVED** [review]
Thanks! Left a couple comments that I hold as blocking approval - one to be actioned before merging, and the other optional in this PR.

**Comment by mdaudali** [comment]
We're about to revert the underlying PR for now, so holding off from merging this in. It will be relevant a little later

**Comment by mdaudali** [comment]
@schlosna - revisiting this, would you be open to rebasing on top of develop so we can get this in?

**Comment by schlosna** [comment]
> @schlosna - revisiting this, would you be open to rebasing on top of develop so we can get this in?

Yeah, I'll rebase and push in the morning.

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosTimestampBoundStore.java
Comment (schlosna) [reply timelock-impl/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosTimestampBoundStore.java]: I'm going to inline these two private methods (each only used once) as my main concern was someone in the future not throwing the returned exception.

**Comment by schlosna** [comment]
Rebased and updated

**Comment by schlosna** [comment]
Will wait to merge & release until Tuesday, May 28th to avoid anything rolling out over the long weekend.

**Pull request merged**
